### PR TITLE
Functional Logging

### PR DESCRIPTION
**Pull request by carmen1234**
## Summary of Changes
Async Logging
wrapper logs 

## Testing Performed
**Environment**: 
plyboy

### CODE REVIEW

**Comment by KyleRAnderson** [comment]
Was this ready for review @carmen1234 ?

File: src/main/java/ca/team2706/frc/robot/Robot.java [hunk 1]
```diff
@@ -159,6 +159,8 @@ public void testPeriodic() {
      * @param args Arguments passed on startup
      */
     public static void main(String[] args) {
+        Logging.init();
```
Comment (KyleRAnderson) [inline src/main/java/ca/team2706/frc/robot/Robot.java]: Should this be called in `robotInit()` or is there a specific reason why it's called here instead?

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package ca.team2706.frc.robot.logging;
+
+import ca.team2706.frc.robot.Robot;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.config.Configurator;
+
+public class Logging {
```
Comment (KyleRAnderson) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: This whole class should probably get some javadoc descriptions for the class as well as all the methods inside of it.

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import ca.team2706.frc.robot.Robot;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.config.Configurator;
+
+/**
+ * Logs to USB and console at levels debug, info, warning, error
+ */
+public class Logging {
```
Comment (eandr127) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: This should be named `Log.java`

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    private static final Logger LOGGER = LogManager.getLogger(Robot.class.getName());
+
+    /**
+     *Starts logging
```
Comment (eandr127) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: There should be a space after the star
```Java
/**
 * Starts logging
 */
```

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * Debug log
+     * @param message The object (or String) message to log
+     */
+    public static void d(Object message) {
+        LOGGER.debug(message);
+    }
+
+    /**
+     *Debug log with exception
```
Comment (eandr127) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: Log formatting

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+     * @param message The object (or String) message to log
+     * @param t The Throwable to log
+     */
+    public static void d(Object message, Throwable t) {
+        LOGGER.debug(message, t);
+    }
+
+    /**
+     *Info log
```
Comment (eandr127) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: Formatting

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+     *Returns the logger
+     * @return logger
+     */
+    public static Logger getLogger() {
+        return LOGGER;
+    }
+
+    /**
+     *Disables logging sets the log level
+     * @param doDisable
```
Comment (eandr127) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: This needs to explain what the parameters do

File: src/main/java/ca/team2706/frc/robot/logging/Logging.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+     * @param message The object (or String) message to log
+     * @param t The Throwable to log
+     */
+    public static void e(Object message, Throwable t) {
+        LOGGER.error(message, t);
+    }
+
+    /**
+     *Returns the logger
+     * @return logger
```
Comment (eandr127) [inline src/main/java/ca/team2706/frc/robot/logging/Logging.java]: Needs to explain what is returned

File: src/main/java/ca/team2706/frc/robot/logging/Log.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        DriverStation.reportWarning((String) message, true);
+    }
+
+    /**
+     * Error log
+     * @param message The object (or String) message to log
+     */
+    public static void e(Object message) {
+        LOGGER.error(message);
+        DriverStation.reportError((String) message, true);
```
Comment (KyleRAnderson) [inline src/main/java/ca/team2706/frc/robot/logging/Log.java]: Wouldn't it make more sense to do `message.toString()`.

File: src/main/java/ca/team2706/frc/robot/logging/Log.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+     * Returns the logger
+     * @return The logger used to log
+     */
+    public static Logger getLogger() {
+        return LOGGER;
+    }
+
+    /**
+     * Disables logging sets the log level
+     * @param doDisable
```
Comment (KyleRAnderson) [inline src/main/java/ca/team2706/frc/robot/logging/Log.java]: Missing description on this parameter.

File: src/main/java/ca/team2706/frc/robot/logging/Log.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * Warning log with exception
+     * @param message The object (or String) message to log
+     * @param t The Throwable to log
+     */
+    public static void w(Object message, Throwable t) {
+        LOGGER.warn(message, t);
+        DriverStation.reportWarning((String) message, true);
```
Comment (KyleRAnderson) [inline src/main/java/ca/team2706/frc/robot/logging/Log.java]: I think it would make more sense to use `message.toString()`.

**Pull request merged**